x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/OnAsyncWaitBranchingStep.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.ilm;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.TriConsumer;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+
+import java.util.Objects;
+
+/**
+ * This step changes its {@link #getNextStepKey()} depending on the result of an {@link AsyncWaitStep}.
+ * <p>
+ * The next step key will point towards the "wait action fulfilled" step when the condition of the async action is met, and towards the
+ * "wait action unfulfilled" step when the client instructs they'd like to stop waiting for the condition using the
+ * {@link BranchingStepListener#onStopWaitingAndMoveToNextKey(ToXContentObject)}.
+ * <p>
+ * If the async action that's branching the execution results in a failure, the {@link #getNextStepKey()} is not set (as ILM would move
+ * into the {@link ErrorStep} step).
+ */
+public class OnAsyncWaitBranchingStep extends AsyncWaitStep {
+    public static final String NAME = "async-branch";
+
+    private static final Logger logger = LogManager.getLogger(OnAsyncWaitBranchingStep.class);
+
+    private StepKey nextStepKeyUnfulfilledWaitAction;
+    private StepKey nextStepKeyFulfilledWaitAction;
+    private TriConsumer<Client, IndexMetaData, BranchingStepListener> asyncWaitAction;
+    private SetOnce<Boolean> onCompleteConditionMet;
+
+    /**
+     * {@link BranchingStep} is a step whose next step is based on the {@link #asyncWaitAction} condition being met or not.
```

Comment:
I think this might be linking to the wrong thing, maybe
```suggestion
     * {@link OnAsyncWaitBranchingStep} is a step whose next step is based on the {@link #asyncWaitAction} condition being met or not.
```